### DISCUSSION TITLE
Slow down with Kotlin versions

### DISCUSSION THREAD

**Discussion by pkubowicz** [LemonAppDev/konsist#1313] [Request a Feature]
### Tell us more.

You are very fast in updating Kotlin version you require. Which makes it hard to use your library.

Let me give you an example. [0.16.0](https://mvnrepository.com/artifact/com.lemonappdev/konsist/0.16.0) requires Kotlin 2.0.20 which was released just 8 days before Konsist.

The whole branch of Kotlin 2 is fairly new. For projects based on Spring ecosystem, it's still an unsupported version. There is no Spring Boot version officially compatible with Kotlin 2. See https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-dependencies/3.3.3

You are also very fast to bump the dependency on Coroutines. If I'm based on [Spring Boot 3.2](https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-dependencies/3.2.9) (still officially getting updates), then I'm not able to use current Konsist because I don't want a test library to mess up with my production dependencies.

Libraries should be conservative in updating its exposed dependencies, because those dependencies are minimum requirements for all downstream. Picking a fresh Kotlin version 8 days after it's made public may be a good policy when you develop an app, but it's not the best practice when you develop a library (and definitely should not happen automatically).

Is there a technical reason you chose to base on Kotlin 2 instead of Kotlin 1.9?

**Comment by igorwojda** [comment]
This is a  very good point.
I will have to rethink the strategy here.

Another option I am considering is usage Konsist as external project - external project would contain Konsist and entire configuration and would point to project being tested.

**Reply by rocketraman** [reply]
You can upgrade to newer versions of Kotlin. But use `apiVersion` and `languageVersion` to make your artifacts compatible with older versions. A good policy to follow would be Kotlin's upstream policy of supporting 3 previous language versions. See https://kotlinlang.org/docs/kotlin-evolution-principles.html#compatibility-options.

**Reply by igorwojda** [reply]
Good tip. 
PR is ready https://github.com/LemonAppDev/konsist/pull/1385
This change will be available with the next release planned for next week.

**Reply by pkubowicz** [reply]
Thank you for the answer!

I would like to point out that me and @rocketraman were talking about different things.
* I was talking about POMs published and the possibility of Konsist library affecting resolution of Kotlin version in Gradle (or Maven if you are unlucky to use it).
* @rocketraman was talking about binary compatibility when calling Konsist from an older version of Kotlin

Also, an *issue mentioned but not addressed is the Coroutines dependency*.

I'm not sure that #1385 is the best solution. I even have doubts that backward binary compatibility is a good thing. To me, a POM expresses a _minimum supported version_, because Gradle makes it easy to use a higher version, but it's neither safe nor easy to use a lower version (you are asking for NoClassDefFoundError). So to me, offering backward binary compatibility is conflicting with requiring a higher version in POM.

I think you should go at least one step ahead.
1. Either put 1.8 not just into bytecode, but also into POM. This is not possible if you need something provided by Kotlin 1.9+. But the plus of this option is that you will avoid a discrepancy, where POM gives a version higher than the bytecode.
2. Or you accept the discrepancy. Then you should clearly document your compatibility with Kotlin *and Coroutines* versions. Because it's not normal that I take a library and don't provide its transitive dependencies (but provide them in lower versions). In normal cases, I will get NoClassDefFoundError. Only for carefully prepared libraries, this has a chance to work. But the library should make it explicit, so that users know downgrading is designed to work, as is not violating library's prerequisites. Without other documentation, your only documentation is the POM.

**Reply by rocketraman** [reply]
@pkubowicz Konsist expresses its dependencies as `implementation` dependencies, as opposed to `api` dependencies:

https://github.com/LemonAppDev/konsist/blob/d9a4fcb4e9427a95cbde079a959984aad78567d4/lib/build.gradle.kts#L7-L9

This means that consumers of the library will not inherit the versions of Kotlin stdlib and coroutines from this library. For the Kotlin stdlib and coroutines, this should always work for up to 3 older versions because of the compatibility guarantees provided by Jetbrains (i.e. they *are* carefully prepared libraries).

Does this address your concerns?

**Reply by pkubowicz** [reply]
> This means that consumers of the library will not inherit the versions of Kotlin stdlib and coroutines from this library.

This is not true.

'implementation' dependencies are not completely hidden. There is no magic classloader separation, no OSGi or stuff. 'implementation' is just 'compile inside the code unit, runtime outside of it'.

When Gradle is publishing to a Maven repository, 'implementation' becomes 'runtime' in POM. Like here: https://mvnrepository.com/artifact/com.lemonappdev/konsist/0.16.1

Here is the simplest demonstration of what it means in practice:

```gradle
plugins {
    kotlin("jvm") version "1.8.0"
}

repositories {
    mavenCentral()
}

dependencies {
    testImplementation("com.lemonappdev:konsist:0.16.1")
}
```

```
% gradle dependencies --configuration=testRuntimeClasspath

testRuntimeClasspath - Runtime classpath of compilation 'test' (target  (jvm)).
+--- org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.8.0 -> 2.0.20
```

Suddenly, my tests are compiled with Kotlin 1.8.0 but executed with kotlin-stdlib 2.0.20. All because of the Konsist dependency.

Maybe nothing bad will happen because Kotlin authors are very careful about backward compat. But it may be dangerous in case of Coroutines dep.

**Reply by rocketraman** [reply]
> 'implementation' dependencies are not completely hidden. There is no magic classloader separation, no OSGi or stuff. 'implementation' is just 'compile inside the code unit, runtime outside of it'.

You're right in terms of runtime dependency, but this is such a common scenario that Gradle provides tooling to avoid it. You can use [Resolution Rules](https://docs.gradle.org/current/userguide/resolution_rules.html) to set your kotlin version. Pretty much all of my projects contain a snippet like this:

```kotlin
resolutionStrategy.eachDependency {
  if (requested.group == "org.jetbrains.kotlin" && requested.name.startsWith("kotlin-")) {
    useVersion(kotlinVersion)
    because("Lock Kotlin artifact versions")
  }
}
```

Furthermore, you mentioned the Spring ecosystem in your original post. Spring does this for you with their dependency-management plugin. For example, if you declare the plugins like the [Spring Boot Kotlin demo](https://github.com/sdeleuze/spring-boot-kotlin-demo/blob/main/build.gradle.kts) does:

```
plugins {
  val kotlinVersion = "2.0.0"
  id("org.springframework.boot") version "3.3.1"
  id("io.spring.dependency-management") version "1.1.5"
  kotlin("jvm") version kotlinVersion
  kotlin("plugin.spring") version kotlinVersion
}

repositories {
    mavenCentral()
}

dependencies {
    testImplementation("com.lemonappdev:konsist:0.16.1")
}
```

And now look at the dependency report:

```
% gradle dependencies --configuration=testRuntimeClasspath
testRuntimeClasspath - Runtime classpath of null/test.
+--- org.jetbrains.kotlin:kotlin-stdlib:2.0.0
...
```

> Maybe nothing bad will happen because Kotlin authors are very careful about backward compat. But it may be dangerous in case of Coroutines dep.

The coroutines dep is also shepherded by Jetbrains, and is also carefully designed and implemented. I don't think you'll have issues. Have you had issues or is this a theoretical discussion?

**Reply by pkubowicz** [reply]
> You can use [Resolution Rules](https://docs.gradle.org/current/userguide/resolution_rules.html) to set your kotlin version

Resolution Rules are for fixing bad POM data: 'One example of a broken module is one that declares a dependency on a library that cannot be found in any of the public repositories'. Greenfield libraries (like this) should not start with broken POM data.

> Spring does this for you with their dependency-management plugin

My main complaint is not that it's hard to downgrade Kotlin. I know more ways to downgrade Kotlin. My complaint is that downgrading is 1) dangerous 2) a failure of the library author because I have to do extra work.

Using a library requiring Kotlin 2.0 with Kotlin 1.8 looks like a bug to me. Unless the library clearly documents that 1.8 is withing the supported version range. With lack of docs, it should be assumed that an older version won't work. If you expect users to correct your POM, explain to them what is within supported ranges.

> The coroutines dep is also shepherded by Jetbrains

I see 'breaking changes' section in https://github.com/Kotlin/kotlinx.coroutines/releases/tag/1.7.0.

**Reply by rocketraman** [reply]
Look, I'm not an author of Konsist, so this will be my last post on the topic. I do think what you are asking for is not reasonable, but at the end of the day I don't have any real say on the matter.

> > You can use [Resolution Rules](https://docs.gradle.org/current/userguide/resolution_rules.html) to set your kotlin version
> 
> Resolution Rules are for fixing bad POM data: 'One example of a broken module is one that declares a dependency on a library that cannot be found in any of the public repositories'. Greenfield libraries (like this) should not start with broken POM data.

Not at all. That is only one use case. There is a whole section of the Gradle docs dedicated to [controlling versions of transitive dependencies](https://docs.gradle.org/current/userguide/dependency_constraints.html) through constraints. "**Customizing Resolution**" is a subtopic of "Controlling Transitives". This functionality exists because it isn't possible to control the downstream versions of every dependency by raising issues with them, as you seem to be trying to do.

Its impossible to define in any case. Your title says "Slow down". Ok, how slow do you want to go? What if their "slow" is still faster than you, or the next guy, wants?

> > Spring does this for you with their dependency-management plugin
> 
> My main complaint is not that it's hard to downgrade Kotlin. I know more ways to downgrade Kotlin. My complaint is that downgrading is 1) dangerous 2) a failure of the library author because I have to do extra work.

This is **always** work you have to do when adding dependencies to a project. Sometimes you have to use an older version of a library. Sometimes you have to update your own deps. Sometimes you have to test the library against the version of the deps you are using. Sometimes you have to fork a downstream dependency. Sometimes you have to structure your project differently (which is particularly easy to do with Konsist!), or break out another service. This is reality.

> Using a library requiring Kotlin 2.0 with Kotlin 1.8 looks like a bug to me. Unless the library clearly documents that 1.8 is withing the supported version range. With lack of docs, it should be assumed that an older version won't work. If you expect users to correct your POM, explain to them what is within supported ranges.

I don't disagree that a project should document its supported ranges. I suspect you won't see any actual problems using it with 1.8. Test it and if it works submit a PR! If it doesn't, submit another PR stating the minimum, or even better modify the code so it does work with the older runtime versions.

> 
> > The coroutines dep is also shepherded by Jetbrains
> 
> I see 'breaking changes' section in https://github.com/Kotlin/kotlinx.coroutines/releases/tag/1.7.0.

None of those breaking changes are likely to affect your usage of Konsist with an older coroutines lib.

**Comment by igorwojda** [comment]
Transitive dependencies can be challenging.

The [semver](https://semver.org/) aims to ensure backward compatibility with `minor` and `patch` updates, but as pointed we have seen breaking changes in `minor` release of coroutine lib, so we can't relay on semver here.

ATK Konsist is using coroutines `1.6.4` derived from `org.jetbrains.kotlin:kotlin-compiler-embeddable`

```
     +--- org.jetbrains.kotlin:kotlin-compiler-embeddable:2.0.20
          ...
     |    \--- org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.6.4
```

The used `1.6.4` version is 3 minor version behind the current one (`1.9.0`) and it was [released on Jul 13, 2022](https://mvnrepository.com/artifact/org.jetbrains.kotlinx/kotlinx-coroutines-core) so I think this should cover most of the projects ATM.
Notice that Konist usage of coroutines is minimal and quite straight forward, so I don't expect breaking changes to affect Konsist in projects using newer versions of coroutines (still we cna take a look if concrete issue occurs).

Also I think Backwards Compatibility https://github.com/LemonAppDev/konsist/pull/1385 will make Konsist work with older versions of  Kotlin.

Keep in mind that major version change (to `Kotlin 2.0`) is an exception, not a rule and it will not happen often. I agree Kotlin 2.0 dependency update was a bit premature, but we will keep it for now to avoid further confusion.

## Proposed solution
- Keep Backwards Compatibility PR trick https://github.com/LemonAppDev/konsist/pull/1385
- Add coroutine version used by Konsist to Konsist docs https://docs.konsist.lemonappdev.com/getting-started/getting-started#add-konsist-dependency
- Freeze automatic Kotlin version upgrades (keep path updates of Kotlin 2.0.x going, but postpone with minor Kotlin releases for much longer) https://github.com/LemonAppDev/konsist/pull/1401/
- Add test projects that tests Konsist with

| Kotlin version  | Coroutine version |
| ------------- | ------------- |
| Current Konsist dependency version  | Current Konsist dependency version  |
| Current Konsist dependency version  | Newest  |
| Min supported from Backwards compatibility  | Current Konsist dependency version   |
| Min supported from Backwards compatibility  | Newest  |

## Future
- Consider adding support for external projects containing Konsist test - this project could verify code present in production project

## Question
@rocketraman @pkubowicz any thing else we can do here to fill the gap? Please propose an ideal solution is you have one in mind.

**Reply by rocketraman** [reply]
Very well thought out and reasonable approach. :+1:

**Reply by igorwojda** [reply]
@pkubowicz any thoughts on this?

**Reply by pkubowicz** [reply]
Hi, thanks for the answer!

I'm happy with the solution. The documentation looks good. Limiting Renovate to patch releases looks like a correct decision as well.

I don't have a big issue with Konsist being based on Kotlin 2.0, as it claims to be backward compatible. Before my question, I could not find declarations from your side on intended Kotlin version support, which made me a bit nervous when developing code based on Konsist. But now I feel much more secure knowing that Konsist won't quickly jump to Kotlin 2.1 or some other bleeding-edge release.

As for 'external projects containing Konsist test', my feedback is that I don't expect to need it.  I'm using a separate Gradle source set to run Konsist. It's a bit complicated to set it up for the first time, but you do it once and don't touch it later. I'm happy with the isolation level offered by the separate source set – this way Konsist classpath is not able to affect execution of my unit tests.

**Comment by nataliapeterwas** [comment]
Konsist `0.17.0` has been released! This release includes this improvement :) See full release log: [Release log](https://github.com/LemonAppDev/konsist/releases/tag/v0.17.0)

**Reply by pkubowicz** [reply]
Thank you!
